src/pool.ts

```diff
@@ -0,0 +1,677 @@
+    if (this._closing) {
+      return PoolState.closing;
+    }
+
+    if (this._closed) {
+      return PoolState.closed;
+    }
+
+    return PoolState.open;
+  }
+
+  constructor(
+    connectOptions: ConnectConfig,
+    minSize: number = DefaultMinPoolSize,
+    maxSize: number = DefaultMaxPoolSize,
+    onAcquire?: (proxy: PoolConnectionProxy) => Promise<void>,
+    onRelease?: (proxy: PoolConnectionProxy) => Promise<void>,
+    onConnect?: (connection: AwaitConnection) => Promise<void>
+  ) {
+    this.validateSizeParameters(minSize, maxSize);
+
+    this._queue = new LifoQueue<PoolConnectionHolder>();
+    this._holders = [];
+    this._initialized = false;
+    this._initializing = false;
+    this._minSize = minSize;
+    this._maxSize = maxSize;
+    this._onAcquire = onAcquire;
+    this._onRelease = onRelease;
+    this._onConnect = onConnect;
+    this._closing = false;
+    this._closed = false;
+    this._generation = 0;
+    this._connectOptions = connectOptions;
+    this._working_addr = null;
+  }
+
+  private validateSizeParameters(minSize: number, maxSize: number): void {
+    if (maxSize <= 0) {
+      throw new InvalidValueError(
+        "maxSize is expected to be greater than zero"
+      );
+    }
+
+    if (minSize < 0) {
+      throw new InvalidValueError(
+        "minSize is expected to be greater or equal to zero"
+      );
+    }
+
+    if (minSize > maxSize) {
+      throw new InvalidValueError("minSize is greater than maxSize");
+    }
+  }
+
+  async initialize(): Promise<void> {
+    // Ref: asyncio_pool.py _async__init__
+    if (this._initialized) {
+      return;
+    }
+    if (this._initializing) {
+      throw new InterfaceError("The pool is already being initialized");
+    }
+    if (this._closed) {
+      throw new InterfaceError("The pool is closed");
+    }
+
+    this._initializing = true;
+
+    for (let i = 0; i < this._maxSize; i++) {
+      const connectionHolder = new PoolConnectionHolder(
+        this,
+        this._onAcquire,
+        this._onRelease
+      );
+
+      this._holders.push(connectionHolder);
+      this._queue.push(connectionHolder);
+    }
+
+    try {
+      await this._initializeHolders();
+    } finally {
+      this._initialized = true;
+      this._initializing = false;
+    }
+  }
+
+  /**
+   * Expires all currently open connections.
+   *
+   * All currently open connections will get replaced on the
+   * next Pool.acquire() call.
+   */
+  expireConnections(): void {
+    // Expire all currently open connections
+    this._generation += 1;
+  }
+
+  private async _initializeHolders(): Promise<void> {
+    if (!this._minSize) {
+      return;
+    }
+
+    // Since we use a LIFO queue, the first items in the queue will be
+    // the last ones in `self._holders`.  We want to pre-connect the
+    // first few connections in the queue, therefore we want to walk
+    // `self._holders` in reverse.
+
+    const reversedHolders = reversed(this._holders);
```

Comment:
I guess there's no need in the `reversed` helper -- just iterate the loop in other direction.